Backport of:

From c020faada2688515f8a7c90ab95f8d5b0b3b82ae Mon Sep 17 00:00:00 2001
From: Nikos Mavrogiannopoulos <nmav@redhat.com>
Date: Mon, 13 Mar 2017 17:00:22 +0100
Subject: [PATCH] Allow reverting the SHA1 ban as a signature algorithm

This allows distributors to decide not to ban SHA1. This
option may be removed in the future.

Signed-off-by: Nikos Mavrogiannopoulos <nmav@redhat.com>
---
 configure.ac         |  1 +
 lib/algorithms/mac.c |  8 +++++++-
 m4/hooks.m4          | 14 ++++++++++++++
 3 files changed, 22 insertions(+), 1 deletion(-)

--- a/configure.ac
+++ b/configure.ac
@@ -942,6 +942,7 @@ AC_MSG_NOTICE([Optional features:
 if features are disabled)
 
   DTLS-SRTP support:    $ac_enable_srtp
+  Allow SHA1 sign:      $ac_allow_sha1
   ALPN support:         $ac_enable_alpn
   OCSP support:         $ac_enable_ocsp
   Ses. ticket support:  $ac_enable_session_tickets
--- a/lib/algorithms/mac.c
+++ b/lib/algorithms/mac.c
@@ -25,8 +25,14 @@
 #include <gnutls_errors.h>
 #include <x509/common.h>
 
+#ifdef ALLOW_SHA1
+# define SHA1_SECURE_VAL 1
+#else
+# define SHA1_SECURE_VAL 0
+#endif
+
 static const mac_entry_st hash_algorithms[] = {
-	{"SHA1", HASH_OID_SHA1, GNUTLS_MAC_SHA1, 20, 20, 0, 0, 0, 64},
+	{"SHA1", HASH_OID_SHA1, GNUTLS_MAC_SHA1, 20, 20, 0, 0, SHA1_SECURE_VAL, 64},
 	{"MD5", HASH_OID_MD5, GNUTLS_MAC_MD5, 16, 16, 0, 0, 0, 64},
 	{"SHA256", HASH_OID_SHA256, GNUTLS_MAC_SHA256, 32, 32, 0, 0, 1,
 	 64},
--- a/m4/hooks.m4
+++ b/m4/hooks.m4
@@ -140,6 +140,20 @@ LIBTASN1_MINIMUM=4.3
     AC_MSG_WARN([C99 macros not supported. This may affect compiling.])
   ])
 
+  ac_allow_sha1=no
+  AC_MSG_CHECKING([whether to allow SHA1 as an acceptable hash for digital signatures])
+  AC_ARG_ENABLE(sha1-support,
+    AS_HELP_STRING([--enable-sha1-support],
+                   [allow SHA1 as an acceptable hash for digital signatures]),
+    ac_allow_sha1=$enableval)
+  if test x$ac_allow_sha1 != xno; then
+   AC_MSG_RESULT(no)
+   AC_DEFINE([ALLOW_SHA1], 1, [allow SHA1 as an acceptable hash for digital signatures])
+  else
+   AC_MSG_RESULT(yes)
+  fi
+  AM_CONDITIONAL(ALLOW_SHA1, test "$ac_allow_sha1" != "no")
+
   ac_enable_srtp=yes
   AC_MSG_CHECKING([whether to disable DTLS-SRTP extension])
   AC_ARG_ENABLE(dtls-srtp-support,
